kernel/src/actions/mod.rs

```diff
@@ -505,6 +505,55 @@ impl Protocol {
     }
 }
 
+// TODO: implement Scalar::From<HashMap<K, V>> so we can derive IntoEngineData using a macro (issue#1083)
+impl IntoEngineData for Protocol {
+    fn into_engine_data(
+        self,
+        schema: SchemaRef,
+        engine: &dyn Engine,
+    ) -> DeltaResult<Box<dyn EngineData>> {
+        let min_reader_version = Scalar::from(self.min_reader_version);
+        let min_writer_version = Scalar::from(self.min_writer_version);
+
+        let reader_features = match self.reader_features {
+            Some(features) => {
+                let features: Vec<Scalar> = features.into_iter().map(Scalar::from).collect();
+                Scalar::Array(ArrayData::try_new(
+                    ArrayType::new(DataType::STRING, false),
+                    features,
+                )?)
+            }
+            None => Scalar::Null(DataType::Array(Box::new(ArrayType::new(
+                DataType::STRING,
+                false,
+            )))),
+        };
+        let writer_features = match self.writer_features {
+            Some(features) => {
+                let features: Vec<Scalar> = features.into_iter().map(Scalar::from).collect();
+                Scalar::Array(ArrayData::try_new(
+                    ArrayType::new(DataType::STRING, false),
+                    features,
+                )?)
+            }
+            None => Scalar::Null(DataType::Array(Box::new(ArrayType::new(
+                DataType::STRING,
+                false,
+            )))),
```

Comment:
Can we de-duplicate these? Perhaps using a generic.
```rust
fn features_to_scalar(features: Option<impl IntoIterator<Item = impl Into<Scalar>>>) -> Scalar {
    match features {
                Some(features) => {
                    let features: Vec<Scalar> = features.into_iter().map(Scalar::from).collect();
                    Scalar::Array(ArrayData::try_new(
                        ArrayType::new(DataType::STRING, false),
                        features,
                    )?)
                }
                None => Scalar::Null(DataType::Array(Box::new(ArrayType::new(
                    DataType::STRING,
                    false,
                )))),
            }
    }
}
```

Reply:
aside: Why did we decide to have `ReaderFeature` and `WriterFeature` instead of a unified `DeltaFeature` with `FeatureType`: `ReaderWriter` or `Writer`

Reply:
great question - probably didn't think too hard about it back then :)

Reply:
Cool. I think this would simplify our feature handling as part of the whole tableconfiguration effort as well. I'll add an issue.